dataplanes/vpp/pkg/nsmvpp/nsm_vpp_memif.go

```diff
@@ -0,0 +1,119 @@
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nsmvpp
+
+import (
+	"fmt"
+	govppapi "git.fd.io/govpp.git/api"
+	"github.com/docker/docker/pkg/mount"
+	"github.com/ligato/networkservicemesh/dataplanes/vpp/pkg/nsmutils"
+	"github.com/sirupsen/logrus"
+	"net"
+	"os"
+	"path"
+)
+
+type parameters map[string]string
+
+const (
+	//NSMSocketFile defines socket name which will be used for memif connection
+	NSMSocketFile = "socketfile"
+	//NSMMaster if true, than role is master
+	NSMMaster = "master"
+	//NSMSlave if true, than role is slave
+	NSMSlave = "slave"
+	//NSMPerPodDirectory defines directory that is mounted to pod (relative to /var/lib/networkservicemesh)
+	NSMPerPodDirectory = "directory"
+
+	BaseDir = "/var/lib/networkservicemesh/"
+
+	MemifDirectory = "/memif"
+)
+
+type MemifInterface struct{}
+
+func (m MemifInterface) ValidateParameters(parameters map[string]string) error {
+	keysList := nsmutils.Keys{
+		NSMSocketFile:      nsmutils.KeyProperties{Validator: nsmutils.Empty},
+		NSMMaster:          nsmutils.KeyProperties{Validator: nsmutils.Bool},
+		NSMSlave:           nsmutils.KeyProperties{Validator: nsmutils.Bool},
+		NSMPerPodDirectory: nsmutils.KeyProperties{Mandatory: true, Validator: nsmutils.Empty},
+	}
+
+	return nsmutils.ValidateParameters(parameters, keysList)
+}
+
+func (m MemifInterface) CreateLocalConnect(apiCh govppapi.Channel, src, dst map[string]string) (string, error) {
+	if err := createMemifSocket(src, dst); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%v-%v", src[NSMPerPodDirectory], dst[NSMPerPodDirectory]), nil
+}
+
+func (m MemifInterface) DeleteLocalConnect(apiCh govppapi.Channel, connID string) error {
+	return nil
+}
+
+func createMemifSocket(src, dst parameters) error {
+	connectionId := buildConnectionId(src, dst)
+
+	if src[NSMSocketFile] == "" && dst[NSMSocketFile] == "" {
+		generatedName := connectionId + ".sock"
+		src[NSMSocketFile] = generatedName
+		dst[NSMSocketFile] = generatedName
+	}
+
+	srcSocketDir, err := createSocketFolder(src, connectionId)
+	if err != nil {
+		return err
+	}
+
+	dstSocketDir, err := createSocketFolder(dst, connectionId)
+	if err != nil {
+		return err
+	}
+
+	if err := mount.Mount(srcSocketDir, dstSocketDir, "hard", "bind"); err != nil {
+		return err
+	}
+	logrus.Infof("Successfully mount folder %s to %s", connectionId, dstSocketDir)
+
+	socket := path.Join(srcSocketDir, src[NSMSocketFile])
+	//dstSocket := path.Join(dstSocketDir, dst[NSMSocketFile])
+
+	if _, err := net.Listen("unix", socket); err != nil {
+		return err
+	}
```

Comment:
Why are we listening on the socket here?

If its a direct memif memif, the NSC/NSE will do so.

If its memif via vpp to memif, then we will need to ask the vpp to do so.

I'm not quite seeing why the vpp-daemon would listen on the socket here (though I'm open to explanation of a reason I don't see :) ).